pkg/cmd/miscellaneous.go

```diff
@@ -213,18 +222,76 @@ func getProjectObject() (*v1beta1.Project, error) {
 	return nil, errors.New("can't determine project")
 }
 
-//getShootObject return shoot object and error
-func getShootObject() (*v1beta1.Shoot, error) {
-	var target Target
-	ReadTarget(pathTarget, &target)
+//GetSeedObject return seed object and error
+func GetSeedObject(targetReader TargetReader, name string) (*v1beta1.Seed, error) {
+	target := targetReader.ReadTarget(pathTarget)
 	gardenClientset, err := target.GardenerClient()
+	if err != nil {
+		return nil, err
+	}
+
+	if name == "" {
+		var shoot *v1beta1.Shoot
+		shoot, err := GetShootObject(targetReader, "")
+		checkError(err)
+		name = *shoot.Spec.SeedName
+		checkError(err)
+	}
+
+	seed, err := gardenClientset.CoreV1beta1().Seeds().Get(name, metav1.GetOptions{})
 	checkError(err)
-	project, err := getProjectObject()
-	checkError(err)
-	shootName, err := getTargetName("shoot")
-	checkError(err)
-	shoot, err := gardenClientset.CoreV1beta1().Shoots(*project.Spec.Namespace).Get(shootName, metav1.GetOptions{})
-	checkError(err)
+	return seed, nil
+}
+
+// GetShootObject fetches shoot object from given target
+func GetShootObject(targetReader TargetReader, name string) (*v1beta1.Shoot, error) {
```

Comment:
I do not understand why you now duplicated the code of the GetProjectObject into this function instead of calling it, like it was done before. Please elaborate what didn't work with the old implementation

Reply:
Do you mean `project, err := gardenClientset.CoreV1beta1().Projects().Get(projectName, metav1.GetOptions{})`  instead of `project,err := GetProjectObject()`  ?

Reply:
after I changed to 

```golang
//GetShootObject return shoot object and error
func GetShootObject(targetReader TargetReader, name string) (*v1beta1.Shoot, error) {
	target := targetReader.ReadTarget(pathTarget)
	var shootName string
	var err error
	if name == "" {
		shootName, err = GetTargetName(targetReader, "shoot")
		checkError(err)
	} else {
		shootName = name
	}
	gardenClientset, err := target.GardenerClient()
	checkError(err)
	project, err := GetTargetedProjectObject(targetReader)
	checkError(err)
	shoot, err := gardenClientset.CoreV1beta1().Shoots(*project.Spec.Namespace).Get(shootName, metav1.GetOptions{})
	checkError(err)
	return shoot, nil
}
```

with testing error

```golang
[It] should pass on get shoot
  /Users/i333878/Work/tedteng/gardenctl/pkg/cmd/get_test.go:196
2020/10/30 14:58:35 Kind:projectnot found from ~/.garden/sessions/plantingSession/target
```

Reply:
Ok I think I found it there is loop in `target - seed -shoot` scenario when I testing. it may the root cause

```
echo "===target===seed===shoot===" 
go run cmd/gardenctl/main.go target --server https://gardener.garden.dev.k8s.ondemand.com --seed aws --shoot gctl-aws > /dev/null
echo "===get===" 
go run cmd/gardenctl/main.go get shoot > /dev/null
```

There is no `project` in target file  so `GetTargetName(targetReader, "project")` not suitable in here. then I change to `GetFromTargetInfo(targetReader, "project")`  the `getTargetMapInfo` inside of `GetFromTargetInfo` also looking for `project` at the first time to initial.  then jump out methods. then loop again and exit at second times when `len(targetInfo) > 0`  . I will check this part continutly at next week. 

```golang
in getTargetmapInfo

if targetInfo["project"] == "" {
		projectObj, err := GetTargetedProjectObject(targetReader)
		checkError(err)
		targetInfo["project"] = projectObj.Name
	}
```

Reply:
```go
func GetTargetedProjectObject(targetReader TargetReader) (*v1beta1.Project, error) {
	name := GetFromTargetInfo(targetReader, "project")
	return GetProjectObject(targetReader, name)
}
```
This is the problem. The functions `GetTargetedXXXXObject` MUST read the name from the target stack of course, that is also what the name suggests.

This is the fix:
```go
func GetShootObject(targetReader TargetReader, name string) (*v1beta1.Shoot, error) {
	target := targetReader.ReadTarget(pathTarget)
	var err error
	if name == "" {
		name, err = GetTargetName(targetReader, "shoot")
		checkError(err)
	}
	gardenClientset, err := target.GardenerClient()
	checkError(err)
	project, err := GetTargetedProjectObject(targetReader)
	checkError(err)
	return gardenClientset.CoreV1beta1().Shoots(*project.Spec.Namespace).Get(name, metav1.GetOptions{})
}

func GetTargetedProjectObject(targetReader TargetReader) (*v1beta1.Project, error) {
	var target Target
	ReadTarget(pathTarget, &target)

	Client, err := target.K8SClientToKind("garden")
	checkError(err)
	if IsTargeted(targetReader, "project") {
		name, err := GetTargetName(targetReader, "project")
		if err != nil {
			return nil, err
		}
		return GetProjectObject(targetReader, name)
	} else if IsTargeted(targetReader, "seed", "shoot") {
		if getRole() == "user" {
			return nil, errors.New("can't determine project. Target project instead of seed")
		}
		seedName, err := GetTargetName(targetReader, "seed")
		checkError(err)
		shootName, err := GetTargetName(targetReader, "shoot")
		checkError(err)

		shootList, err := Client.GardenerV1beta1().Shoots(metav1.NamespaceAll).List(metav1.ListOptions{
			FieldSelector: fields.SelectorFromSet(
				fields.Set{
					core.ShootSeedName: seedName,
					"metadata.name":    shootName,
				}).String(),
		})
		if err != nil {
			return nil, err
		}
		if len(shootList.Items) == 0 {
			return nil, errors.New("No shoot found with name " + shootName + " that is running on the seed " + seedName)
		} else if len(shootList.Items) > 1 {
			return nil, errors.New("There are multiple shoots with the name " + shootName + " that are running on the seed " + seedName)
		}
		projectNamespace := shootList.Items[0].Namespace

		projectList, err := Client.GardenerV1beta1().Projects().List(metav1.ListOptions{})
		if err != nil {
			return nil, err
		}

		for _, p := range projectList.Items {
			if *p.Spec.Namespace == projectNamespace {
				return &p, nil
			}
		}
	}

	return nil, errors.New("can't determine project")
}

func GetProjectObject(targetReader TargetReader, name string) (*v1beta1.Project, error) {
	var err error
	if name == "" {
		name, err = GetTargetName(targetReader, "project")
		if err != nil {
			return nil, err
		}
	}
	return Client.GardenerV1beta1().Projects().Get(name, metav1.GetOptions{})
}
```

Reply:
It seems with a new error when I check `ci/test` . I did a little bit change. it all fix now.

Reply:
somehow us not able to compatible with existing test suit.

![image](https://user-images.githubusercontent.com/42234376/97830595-708a7c80-1d08-11eb-9004-47ba6da758ad.png)

Reply:
I did a little bit change but not use methods as below , maybe we can fix in the future. 
```golang
// GetShootObject return shoot object and error
func GetShootObject(targetReader TargetReader, name string) (*v1beta1.Shoot, error) {
	target := targetReader.ReadTarget(pathTarget)

	var err error
	if name == "" {
		name, err = GetTargetName(targetReader, "shoot")
		checkError(err)
	}
	gardenClientset, err := target.GardenerClient()
	checkError(err)
	project, err := GetTargetedProjectObject(targetReader)
	checkError(err)

	return gardenClientset.CoreV1beta1().Shoots(*project.Spec.Namespace).Get(name, metav1.GetOptions{})
}
```

Reply:
Test result

```bash
Ran 61 of 61 Specs in 1.040 seconds
SUCCESS! -- 61 Passed | 0 Failed | 0 Pending | 0 Skipped
PASS


+ echo ===target===seed===shoot===
===target===seed===shoot===
+ go run cmd/gardenctl/main.go target --server https://gardener.garden.dev.k8s.ondemand.com --seed aws --shoot gctl-aws
+ check
+ '[' 0 -eq 0 ']'
+ echo PASS
PASS
+ echo ===get===
===get===
+ check
+ '[' 0 -eq 0 ']'
+ echo PASS
PASS
+ go run cmd/gardenctl/main.go get garden
+ check
+ '[' 0 -eq 0 ']'
+ echo PASS
PASS
+ go run cmd/gardenctl/main.go get shoot
+ check
+ '[' 0 -eq 0 ']'
+ echo PASS
PASS
+ go run cmd/gardenctl/main.go get seed
+ check
+ '[' 0 -eq 0 ']'
+ echo PASS
PASS

===get===project===shoot===
+ go run cmd/gardenctl/main.go get garden
+ check
+ '[' 0 -eq 0 ']'
+ echo PASS
PASS
+ go run cmd/gardenctl/main.go get project
+ check
+ '[' 0 -eq 0 ']'
+ echo PASS
PASS
+ go run cmd/gardenctl/main.go get shoot
+ check
+ '[' 0 -eq 0 ']'
+ echo PASS
PASS
+ go run cmd/gardenctl/main.go get seed
+ check
+ '[' 0 -eq 0 ']'
+ echo PASS
PASS
```

Reply:
Hi @tedteng,
I fixed
```go
func GetTargetedProjectObject(targetReader TargetReader) (*v1beta1.Project, error) {
	// var target Target - UPDATE: removed
	// ReadTarget(pathTarget, &target) - UPDATE: removed
	target := targetReader.ReadTarget(pathTarget)
```

then in get_test.go, see my comments at the end of the line... AnyTimes is missing
```go
It("should pass on get shoot", func() {
  // ... 
  target.EXPECT().K8SClientToKind(cmd.TargetKindGarden).Return(k8sClientToGarden, nil).AnyTimes() // <----- AnyTimes missing
  target.EXPECT().K8SClientToKind(cmd.TargetKindSeed).Return(k8sClientToGarden, nil).AnyTimes() // <----- AnyTimes missing
```

probably you also need to adapt the getRole function and pass the targetReader
```go
func getRole(targetReader TargetReader) string {
```

please try the tests again with those fixes. If there are problems with the tests we should fix it, instead of working around it in the productive code

Reply:
In fact, I was also noticed `.AnyTimes()` when I testing, which pop up from the error msg.  but I can't remember whether I change here , because too many verify and change from my local try to passing the testing.  I will check it again. 

```golang
func GetTargetedProjectObject(targetReader TargetReader) (*v1beta1.Project, error) {
	// var target Target - UPDATE: removed
	// ReadTarget(pathTarget, &target) - UPDATE: removed
	target := targetReader.ReadTarget(pathTarget)
```

Reply:
with the latest change today.  I met one new error. 

```golang
[It] should pass on get shoot
  /Users/i333878/Work/tedteng/gardenctl/pkg/cmd/get_test.go:197
2020/11/03 14:18:52 no kind "ShootList" is registered for version "core.gardener.cloud/v1beta1" in scheme "k8s.io/client-go/kubernetes/fake/register.go:67"
```

during the debug, it seems ShootList not get any Items when SelectorFromSet `core.ShootSeedName: seedName, "metadata.name":    shootName,` , I am feeling maybe something in mock, but I couldn't find anything wrong from the mock. Do you know is anything I missing. Thanks.

```golang
			clientSet := gardencorefake.NewSimpleClientset(
				&gardencorev1beta1.Seed{
					ObjectMeta: metav1.ObjectMeta{
						Name: seedName,
					},
					Spec: gardencorev1beta1.SeedSpec{
						SecretRef: &corev1.SecretReference{
							Name:      "test-secret-name",
							Namespace: "test-namespace",
						},
					},
				},
				&gardencorev1beta1.Shoot{
					ObjectMeta: metav1.ObjectMeta{
						Name: "test-shoot",
					},
					Spec: gardencorev1beta1.ShootSpec{
						SeedName: &seedName,
					},
				},
			)
```


![image](https://user-images.githubusercontent.com/42234376/97956094-c2073a00-1de2-11eb-87e0-e921b00b8446.png)

Reply:
I guess it's missing a fake for ShootList
```go
clientSet := gardencorefake.NewSimpleClientset(
  &gardencorev1beta1.Seed{
    // ...
  },
  &gardencorev1beta1.Shoot{
    // ...
  },
  &gardencorev1beta1.ShootList{
    ListMeta: // ... ,
    Items:  // ...  ,
  }
)
```

Reply:
I am not familiar gomock here,  If I understanding correctly the test case in here is getting shoot when target  -s seed -t shoot. Thanks, point the `gardencorev1beta1.ShootList` I will take a look how to mock the data.

```golang
			It("should pass on get shoot", func() {
				targetReader.EXPECT().ReadTarget(gomock.Any()).Return(target).AnyTimes()
				target.EXPECT().K8SClientToKind(cmd.TargetKindGarden).Return(k8sClientToGarden, nil).AnyTimes()
				target.EXPECT().K8SClientToKind(cmd.TargetKindSeed).Return(k8sClientToGarden, nil).AnyTimes()
				target.EXPECT().Stack().Return(targetMeta).AnyTimes()
				target.EXPECT().GardenerClient().Return(clientSet, nil).AnyTimes()
				kubeconfigWriter.EXPECT().Write(gomock.Any(), gomock.Any()).Return(nil)

				ioStreams, _, _, _ := cmd.NewTestIOStreams()
				command = cmd.NewGetCmd(targetReader, configReader, kubeconfigReader, kubeconfigWriter, ioStreams)
				command.SetArgs([]string{"shoot"})
				err := command.Execute()

				Expect(err).NotTo(HaveOccurred())
			})
```

Reply:
```golang
no kind "ShootList" is registered for version "core.gardener.cloud/v1beta1" in scheme "k8s.io/client-go/kubernetes/fake/register.go:67"
```
after a series of testing, it seems issue related within file  https://github.com/gardener/gardenctl/blob/master/vendor/k8s.io/client-go/kubernetes/fake/register.go, however which is beyond my knowledge so far. 

I have to switch my previous methods which able to passing the go test.

Reply:
I just checked again and adding the shootList to the fake is probably not necessary I guess.
The reason for the `no kind "ShootList" is registered` is that the client is using a scheme where the gardener types are not registered.
When using `gardenClientset, err := target.GardenerClient()` instead of `target.K8SClientToKind("garden")` this error does not occur.

Reply:
nice. Thanks, that's works after I add more fake data in clientSet.